File: app/src/main/java/com/shar2wy/moviesapp/models/trailerRepo/TrailerRepo.java
package com.shar2wy.moviesapp.models.trailerRepo;

import android.content.Context;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.functions.Consumer;

/**
 * Created by shar2wy
 * on 4/4/18.
 * Description: description goes here
 */
public class TrailerRepo {

    private static TrailerRepo INSTANCE;
    private TrailerRemoteDataSource remoteDataSource;
    private TrailerLocalDataSource localDataSource;

    private TrailerRepo(Context context) {
        this.remoteDataSource = new TrailerRemoteDataSource(context);
        this.localDataSource = new TrailerLocalDataSource();
    }

    public static synchronized TrailerRepo getInstance(Context context) {
        if (INSTANCE == null) {
            INSTANCE = new TrailerRepo(context);
        }
        return INSTANCE;
    }

    public Flowable<List<Trailer>> getTrailers(int id) {
        return Flowable.concat(localDataSource.getTrailers(id),
                remoteDataSource.getTrailers(id))
                .doOnNext(new Consumer<List<Trailer>>() {
                    @Override
                    public void accept(List<Trailer> trailers) throws Exception {
                        localDataSource.saveTrailers(trailers);
                    }
                });
    }
}

File: app/src/main/java/com/shar2wy/moviesapp/models/trailerRepo/TrailerRepo.kt
package com.shar2wy.moviesapp.models.trailerRepo

import io.reactivex.Flowable

/**
 * Created by shar2wy
 * on 4/4/18.
 * Description: description goes here
 */
class TrailerRepo private constructor() {
    private val remoteDataSource: TrailerRemoteDataSource = TrailerRemoteDataSource()
    private val localDataSource: TrailerLocalDataSource = TrailerLocalDataSource()

    fun getTrailers(id: Int): Flowable<List<Trailer>> {
        return Flowable.concat(localDataSource.getTrailers(id),
                remoteDataSource.getTrailers(id))
                .doOnNext { trailers -> localDataSource.saveTrailers(trailers) }
    }

    companion object {

        private var INSTANCE: TrailerRepo? = null

        @Synchronized
        fun getInstance(): TrailerRepo {
            if (INSTANCE == null) {
                INSTANCE = TrailerRepo()
            }
            return INSTANCE as TrailerRepo
        }
    }
}
